### PR TITLE
Add Quarkus Hazelcast Client extension

### PR DESCRIPTION
**Pull request by pivovarit**
The `1.0.0-RC` version is not released yet (actually, there are no versions released yet). Dropping it to obtain early feedback.

http://github.com/hazelcast/quarkus-hazelcast-client

### CODE REVIEW

**Comment by maxandersen** [comment]
on first look in context on how it technical is done in the notion of universe platform it looks good. your metadata is showing up right.

we really need to get cross integration testing working so we can maintain/ensure external extensions works.

**Comment by geoand** [comment]
We are now close with the integration tests :)

**Comment by pivovarit** [comment]
Great :)

We're dealing with one more issue(https://github.com/hazelcast/quarkus-hazelcast-client/pull/20), release the new version and finish integration here.

**Comment by maxandersen** [comment]
Hey, would you be up for setting up ecosystem-ci for your hazelcast extension ? see this thread at https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/quarkus-dev/xbAARRZC_JU/HseM-bunBQAJ for details.

**Comment by pivovarit** [comment]
note to self:

> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.0:test (default-test) on project kogito-quarkus-integration-test: There are test failures.
[ERROR]

**Comment by pivovarit** [comment]
@maxandersen @geoand I'm on it. Just one question. We're ready to release the next RC which is likely to become GA. 

Is there any value in introducing RC version to quarkus-platform and shall we wait until we make it GA?

**Comment by geoand** [comment]
I personally think there is. @gsmet WDYT?

**Comment by maxandersen** [comment]
I would say yes - we would mark it with "preview" anyways in the platform which allow us to remove/change things. 

Something we will probably do with external extensions until we get the external catalog/registry feature fully baked in.

Does that make sense @pivovarit ?

btw. the ecosystem-ci at the moment does not require inclusion in the platform - it is just fine and even encouraged to do if you have something fully outside.

**Comment by pivovarit** [comment]
Yeah, works for us - it'd be more than reasonable to introduce it as _preview_ at this stage.

I'm trying now to play around with ecosystem-ci and you can expect PRs soon

**Comment by pivovarit** [comment]
We've just released `1.0.0-RC3` and as soon as it propagates to Central, we'll be ready to move on.

**Comment by pivovarit** [comment]
I think we're ready for review here

**Comment by geoand** [comment]
Sorry for letting this slip (which results in having conflicts), but we had to release version `1.4` of the platform.
Can you please rebase so we can then get this in hopefully soon?

**Comment by pivovarit** [comment]
Sure, will do that now!

**Comment by geoand** [comment]
Thanks!

**Comment by pivovarit** [comment]
Looks green already :)

**Comment by geoand** [comment]
Excellent! @gsmet @maxandersen can we merge this?

File: bom/deployment/pom.xml [hunk 1]
```diff
@@ -58,6 +58,12 @@
                 <scope>import</scope>
             </dependency>
 
+            <!-- Hazelcast Client -->
+            <dependency>
+                <groupId>com.hazelcast</groupId>
+                <artifactId>quarkus-hazelcast-deployment</artifactId>
+                <version>${quarkus-hazelcast-client.version}</version>
```
Comment (gsmet) [inline bom/deployment/pom.xml:65]: Better put that one before the Camel one. See the comment of the Camel one.

File: integration-tests/hazelcast-client/src/test/resources/classpathseed [hunk 2]
```diff
@@ -0,0 +1,4 @@
+The purpose of this file is to have a non-empty test-classes directory that will
+be added to the test classpath and used by the Quarkus test extension for storing
+generated classes and other resources that should appear on the classpath during
+the test execution.
```
Comment (gsmet) [inline integration-tests/hazelcast-client/src/test/resources/classpathseed:4]: I don't think this file is needed anymore. @ppalaga can you confirm?

**Comment by gsmet** [comment]
@pivovarit have you checked your extension is properly included in the bom and the json file? (the one generated in the bom/runtime module)

**Comment by geoand** [comment]
Yeah, we absolutely need a squash 😎

**Comment by gsmet** [comment]
@geoand is the ecosystem CI set up? For me, it's a prerequisite now.

**Comment by geoand** [comment]
@gsmet yes, it runs every night: https://github.com/quarkusio/quarkus-platform/actions

**Comment by geoand** [comment]
Testing script is here: https://github.com/quarkusio/quarkus-platform/blob/master/.github/quarkus-ecosystem-test

**Comment by pivovarit** [comment]
I believe I read somewhere that the CI setup needs to happen after the extension gets merged, but I can do it already. The Github Actions pipeline is merged already and I just need to finish setting up the issue and envs.

Yes, I verified that the extension is included in the JSON file:
```
 {
            "name": "Hazelcast Client",
            "metadata": {
                "keywords": [
                    "hazelcast-client",
                    "hazelcast",
                    "cache",
                    "distributed cache",
                    "caching"
                ],
                "guide": "https://quarkus.io/guides/hazelcast-client",
                "categories": [
                    "data"
                ],
                "status": "preview"
            },
            "group-id": "com.hazelcast",
            "artifact-id": "quarkus-hazelcast-client",
            "version": "1.0.0-RC3",
            "description": "Quarkus Hazelcast Client Extension"
        }
```

Also, can't we just use GitHub's Squash'n'Merge?

**Comment by geoand** [comment]
We can't squash the commits from the Github UI because we have elected to always have a merge commit. With that type of merge GithHub UI doesn't provide the ability to squash as well.

File: integration-tests/hazelcast-client/src/test/resources/classpathseed (hunk 2 above)
Comment (pivovarit) [reply integration-tests/hazelcast-client/src/test/resources/classpathseed:4]: FYI I added it after having a look at `debezium-outbox` and `qpid-jms`, which still have these

**Comment by maxandersen** [comment]
> I believe I read somewhere that the CI setup needs to happen after the extension gets merged, but I can do it already.

if that is written somewhere let us know as that is not true. we do NOT need every extension in the platform - any extension interested in keeping compatible and give feedback to Quarkus should add themself to ecoystem-ci. No requirement for platform inclusion :)

**Comment by pivovarit** [comment]
@maxandersen I went through the publicly available docs and I see no traces of it. It might have been here somewhere in the purged comments after a rebase. 

I will let you know if I find something!

**Comment by pivovarit** [comment]
@geoand CI set up (hopefully) and commits squashed

File: integration-tests/hazelcast-client/src/test/resources/classpathseed (hunk 2 above)
Comment (ppalaga) [reply integration-tests/hazelcast-client/src/test/resources/classpathseed:4]: Yes, they do not seem to be needed anymore, @aloubyansky may want to confirm. Just try to remove them and check whether the tests are both executed and passing.

Comment (aloubyansky) [reply integration-tests/hazelcast-client/src/test/resources/classpathseed:4]: I think it still is. It can easily be verified.

Comment (pivovarit) [reply integration-tests/hazelcast-client/src/test/resources/classpathseed:4]: Just to double-check... what would be the definition of "it works"?

We remove the file, run the build, and see if tests from the `*-integration-test` artifact run? or should I look for some generated test sources somewhere?

Comment (gsmet) [reply integration-tests/hazelcast-client/src/test/resources/classpathseed:4]: Not worth the hassle, let's keep it if nobody is sure.

**Comment by gsmet** [comment]
@pivovarit merged, thanks for this work and your patience/tenacity!

**Comment by pivovarit** [comment]
Thank you for the assistance!

**Comment by pivovarit** [comment]
What else do we need to do to make the extension visible on https://code.quarkus.io?

**Comment by gsmet** [comment]
Nothing in theory but I will check tomorrow if everything is ok. It will be part of the Quarkus 1.5 release!

I didn’t want to take the risk to include new things in the bom in a micro.

**Comment by gsmet** [comment]
@pivovarit OK, I confirm it will appear.

Maybe the description could use some love? It says `Quarkus Hazelcast Client Extension` which is a bit redundant.

We tried to word things starting with a verb: "Connect to the MariaDB database via JDBC" for instance.

Maybe "Connect to a Hazelcast cache cluster" or something similar? The description usually comes from the Maven description of your runtime artifact if you haven't modified it in the yaml.

Obviously, this is not blocking :).

**Comment by pivovarit** [comment]
Sure, let me take care of that :)

**Pull request merged**